### PR TITLE
Register tables

### PR DESCRIPTION
**Pull request by ThomasHepworth**
Initial draft for table registration which addresses [issue #585](https://github.com/moj-analytical-services/splink/issues/585).
<br>

**This PR:**
* Add a table registration method to each linker.
* Adds a `query_sql` method for each linker, which allows users to easily query their database.
* Updates our tests to reflect this API change and add some new tests to check that table registration is working as intended.

<hr>

#### Table Registration

Fairly straightforward, but this grants users the ability to simply pass either: a dictionary, record level dictionary, pandas df or pyarrow in the duckdb case/spark df in the spark case.

This is done simply by calling:
`linker.register_table(<input>, <table_name>)`

**To note:** pyarrow is significantly quicker than pandas for this. From some quick benchmarks, I was finding that registering a record level dictionary of 3 million cells into duckdb took:
* 7 seconds with pandas
* 2 seconds with pyarrow.

If we're happy to let pyarrow become a dependency, I'll adjust this PR to use pyarrow instead.

### Query SQL

Not much to add - allows user to query their database as they please.
The syntax is simply: `linker.query_sql(sql)` and will output a pandas df.

### CODE REVIEW

**Comment by ThomasHepworth** [comment]
A couple of extra notes on this work.

It might make sense to add an "accepted data types" list for each linker and check against this whenever users try to register tables.

If an invalid type is passed by the user, we error and give them an informative printout stating that that data type is not supported for that specific linker variant.

I've also not added support for **pyarrow** on any linker other than **duckdb** (where it's baked in). This is because it would require a **pyarrow** dependency in our poetry.lock file, which I was trying to avoid.

File: splink/linker.py [hunk 1]
```diff
@@ -378,6 +378,43 @@ def _execute_sql_against_backend(self, sql, templated_name, physical_name):
             f"_execute_sql_against_backend not implemented for {type(self)}"
         )
 
+    def register_table(self, input, table_name):
+        """
```
Comment (RobinL) [inline splink/linker.py:382]: Need to add this to the docs [here](https://github.com/moj-analytical-services/splink/blob/65586d05a6523ae1f6535d85d79fa65cef75ab0a/docs/linker.md#L40)

File: splink/linker.py [hunk 2]
```diff
@@ -378,6 +378,43 @@ def _execute_sql_against_backend(self, sql, templated_name, physical_name):
...
+        Args:
+            input: The data you wish to register. This can be either a dictionary,
+                pandas dataframe, pyarrow table or a spark dataframe.
+            table_name (str): The name you wish to assign to the table.
+        """
+
+        raise NotImplementedError(f"register_table not implemented for {type(self)}")
+
+    def query_sql(self, sql):
+        """
```
Comment (RobinL) [inline splink/linker.py:403]: Again, need to add to docs

File: splink/spark/spark_linker.py [hunk 3]
```diff
@@ -78,36 +80,36 @@ def __init__(
...
-                Spark Dataframes.
-            repartition_after_blocking (bool, optional): In some cases, especially when
-                the comparisons are very computationally intensive, performance may be
-                improved by repartitioning after blocking to distribute the workload of
-                computing the comparison vectors more evenly and reduce the number of
-                tasks. Defaults to False.
-            num_partitions_on_repartition (int, optional): When saving out intermediate
-                results, how many partitions to use?  This should be set so that
-                partitions are roughly 100Mb. Defaults to 100.
+                holds the data linkage model.
```
Comment (RobinL) [inline splink/spark/spark_linker.py:83]: Is this intentionally indentned?  I don't think it needs to be?

File: tests/test_full_example_spark.py [hunk 4]
```diff
@@ -1,3 +1,4 @@
+from atexit import register
```
Comment (RobinL) [inline tests/test_full_example_spark.py:1]: is this an error introduced by vs code/autoimport?

**Review by RobinL — CHANGES_REQUESTED** [review]
Excellent work!  Some minor comments but looks great to me.

Once you've addressed the comments, add another commit that bump the version (in both __init__ and pyproject.toml) and do a release :-)

File: splink/linker.py [hunk 5]
```diff
@@ -378,6 +378,43 @@ def _execute_sql_against_backend(self, sql, templated_name, physical_name):
...
+
+        Args:
+            input: The data you wish to register. This can be either a dictionary,
+                pandas dataframe, pyarrow table or a spark dataframe.
+            table_name (str): The name you wish to assign to the table.
+        """
+
+        raise NotImplementedError(f"register_table not implemented for {type(self)}")
+
+    def query_sql(self, sql):
```
Comment (RobinL) [inline splink/linker.py:402]: DO we definitely want this to return a pandas dataframe?  

I'm not sure that's such a good idea with big data so I think probably it should return a `SplinkDataframe`, which would also be more consistent with the rest of the API.

(Sorry if im missing something obvious)

File: splink/spark/spark_linker.py (hunk 3 above)
Comment (ThomasHepworth) [reply splink/spark/spark_linker.py:83]: Not a clue what's happened there 🤔. It'll be reverted after the next commit.

**Comment by RobinL** [comment]
Previously:

- The SplinkDataFrame abstraction represented a table in the underlying database
- Methods were provided to turn this table into a pandas dataframe, a spark dataframe, a list-of-record-dicts etc.

- The main linker.py has a '_sql_to_splink_dataframe_checking_cache' that hashed the SQL and retrieved from cache if existed

If we want to execute arbitrary SQL on top of a splink dataframe, the problem is that the existing function `_execute_sql_against_backend` would materialise the results in the and return a SplinkDataFrame

We want to be able to execute arbitrary SQL on top of a SplinkDataFrame and return either:
- A splink dataframe, (using `_execute_sql_against_backend`)
- A pandas dataframe, (here we _don't_ want to use '_execute_sql_against_backend' method, because that method doesn't give you control over materalisation of table)

**Comment by RobinL** [comment]
I think for the sake of simplicity I'd probably not bother with the new functionality you've introduced - i.e. ditch the ability to run a SQL query without the results being persisted to the DB.  

I'm not sure the new query function will be used often enough to justify this extra complexity

But if you want to introduce it, possibly either we need to 
- Modify `_execute_sql_against_backend` so that optionally (using an argument) it can return a pandas or arrow dataframe
- Have a separate function implemented against each backend-specific linker that does this.

Or perhaps there could be a lazy version of the `SplinkDataFrame`.

**Comment by RobinL** [comment]
Another simple option could be to have the `def query_sql(self, sql):` delete the table if the user asks for a pandas dataframe.

I don't think I'd bother supporting lots of output types.  Just pandas and as a splink dataframe.  If the user wants to do something complex, they can always go via the splink dataframe

**Review by RobinL — APPROVED** [review]
I think looks good now once you've fixed linting 🚀

**Comment by ThomasHepworth** [comment]
> 

Just tweaking athena very quickly -> linting and then I'll merge it.

Are you happy w/ me bumping splink to v3.4.0?

**Comment by RobinL** [comment]
Sounds good to me

**Pull request merged**